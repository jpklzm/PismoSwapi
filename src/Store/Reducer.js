const Reducer = (state, action) => {
  switch (action.type) {
    case "SET_CURRENT_MOVIE":
      return {
        ...state,
        movie: action.payload,
      };
    case "SET_CURRENT_MOVIE_LIST":
      return {
        ...state,
        movieList: action.payload,
      };
    default:
      return state;
  }
};

export default Reducer;
