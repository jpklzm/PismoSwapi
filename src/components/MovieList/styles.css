.movie-container {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  flex-direction: column;
  align-items: center;
  padding: 30px;
}

.title {
  color: white;
  margin: auto;
  width: 50%;
  padding: 10px;
  text-align: center;
  font-size: 50px;
}

.red {
  margin: 0;
  color: red;
  border-bottom: solid red;
}

.blue {
  margin: 0;
  color: blue;
  border-bottom: solid blue;
}
