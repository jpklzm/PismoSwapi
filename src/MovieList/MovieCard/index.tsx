import React, { FunctionComponent, useContext, useCallback } from "react";
import "./styles.css";
import { Context } from "../../Store/Store";
import { useHistory } from "react-router-dom";

type MovieCardProps = {
  movie: { title: string };
};

const MovieCard: FunctionComponent<MovieCardProps> = ({ movie }) => {
  //@ts-ignore
  const [state, dispatch] = useContext(Context);

  const history = useHistory();
  const handleOnClick = useCallback(
    (currentMovie) => {
      dispatch({ type: "SET_CURRENT_MOVIE", payload: currentMovie });

      history.push("/movie-details");
    },
    [history]
  );

  return (
    <div>
      {movie && (
        <>
          <div
            className="movie"
            onClick={() => {
              handleOnClick(movie);
            }}
          >
            <h2>{movie.title}</h2>
          </div>
        </>
      )}
    </div>
  );
};

export default MovieCard;
