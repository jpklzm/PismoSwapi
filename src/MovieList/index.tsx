import React, { useState, useEffect } from "react";
import "./styles.css";
import MovieCard from "./MovieCard";

const MovieList = () => {
  const [movieList, setMovieList] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    fetch(`https://swapi.dev/api/films`, {
      method: "GET",
    })
      .then((res) => res.json())
      .then((response) => {
        setMovieList(response.results);
        setIsLoading(false);
      })
      .catch((error) => {
        console.log(error);
      });
  }, []);

  console.log(movieList);
  const mock = [
    {
      title: "um filme",
    },
    {
      title: "outro filme",
    },
  ];

  return (
    <div>
      <h1 className="title">
        <span className="red">Pismo</span> <span className="blue">SWAPI</span>
      </h1>
      <div className="movie-container">
        {isLoading && (
          <img src="https://64.media.tumblr.com/c0d8be3a1d2bfd58a1eb6c91baa5747b/5321eccd302337fc-fc/s400x600/f5ceaa1e1f5a5f705142e087ed9369c0eff52c63.gifv" />
        )}

        {(mock as any).map((movie: any, index: any) => (
          <MovieCard key={index} movie={movie} />
        ))}

        {/* {(movieList as any).map((movie: any, index: any) => (
          <MovieCard key={index} movie={movie} />
        ))} */}
      </div>
    </div>
  );
};

export default MovieList;
