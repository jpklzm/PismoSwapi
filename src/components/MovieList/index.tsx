import React, { useState, useEffect, useContext } from "react";
import "./styles.css";
import MovieCard from "./MovieCard";
import { Context } from "../Store/Store";

const MovieList = () => {
  const [movieList, setMovieList] = useState([]);
  const [filteredList, setFilteredList] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  //@ts-ignore
  const [state, dispatch] = useContext(Context);

  useEffect(() => {
    if (state.movieList != {}) {
      fetch(`https://swapi.dev/api/films`, {
        method: "GET",
      })
        .then((res) => res.json())
        .then((response) => {
          setMovieList(response.results);
          setFilteredList(response.results);
          dispatch({
            type: "SET_CURRENT_MOVIE_LIST",
            payload: response.results,
          });
          setIsLoading(false);
        })
        .catch((error) => {
          console.log(error);
        });
    } else {
      setMovieList(state.movieList);
      setFilteredList(state.movieList);
    }
  }, []);

  const countCharacters = () => {
    const characterCount = [
      {
        url: "http://swapi.dev/api/people/1/",
        count: 0,
      },
    ];
    movieList.forEach((movie: any) => {
      movie.characters.forEach((e: any) => {
        const character = characterCount.find((c: any) => c.url === e);
        if (character) {
          const index = characterCount.findIndex((ch: any) => {
            return ch.url === character.url;
          });
          characterCount[index].count += 1;
        } else {
          characterCount.push({
            url: e.url,
            count: 0,
          });
        }
      });
    });

    console.log(characterCount);
  };

  const clearFilter = () => {
    setFilteredList(state.movieList);
  };

  const filterByCharacter = (character: string) => {
    const filteredMovieList: any = [];

    movieList.forEach((movie: any) => {
      movie.characters.forEach((e: any) => {
        if (e === character) {
          filteredMovieList.push(movie);
        }
      });
    });

    setFilteredList(filteredMovieList);
  };

  return (
    <div>
      <h1 className="title">
        <span className="red">Pismo</span> <span className="blue">SWAPI</span>
      </h1>
      <div className="movie-container">
        {isLoading && (
          <img src="https://64.media.tumblr.com/c0d8be3a1d2bfd58a1eb6c91baa5747b/5321eccd302337fc-fc/s400x600/f5ceaa1e1f5a5f705142e087ed9369c0eff52c63.gifv" />
        )}

        <button
          onClick={() => {
            clearFilter();
          }}
        >
          Clear
        </button>
        <button
          onClick={() => {
            filterByCharacter("http://swapi.dev/api/people/1/");
          }}
        >
          Luke Skywalker
        </button>
        <button
          onClick={() => {
            countCharacters();
          }}
        >
          Count
        </button>

        {(filteredList as any).map((movie: any, index: any) => (
          <MovieCard key={index} movie={movie} />
        ))}
      </div>
    </div>
  );
};

export default MovieList;
