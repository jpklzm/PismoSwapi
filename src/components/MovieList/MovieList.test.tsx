import { mount } from "enzyme";
import MovieList from "./index";
import { Context } from "../Store/Store";

const contextValues = {
  movieList: [{ title: "filme" }, { title: "outrofilme" }],
};

describe("MovieList", () => {
  it("renders correctly", () => {
    const wrapper = mount(
      <Context.Provider value={[contextValues, ""]}>
        <MovieList />
      </Context.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
