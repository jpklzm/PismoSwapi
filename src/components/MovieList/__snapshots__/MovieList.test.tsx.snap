// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList renders correctly 1`] = `
<MovieList>
  <div>
    <h1
      className="title"
    >
      <span
        className="red"
      >
        Pismo
      </span>
       
      <span
        className="blue"
      >
        SWAPI
      </span>
    </h1>
    <div
      className="movie-container"
    >
      <img
        src="https://64.media.tumblr.com/c0d8be3a1d2bfd58a1eb6c91baa5747b/5321eccd302337fc-fc/s400x600/f5ceaa1e1f5a5f705142e087ed9369c0eff52c63.gifv"
      />
    </div>
  </div>
</MovieList>
`;
